src/core/credrank/nodeGadgets.js

```diff
@@ -0,0 +1,104 @@
+// @flow
+/**
+ * This module contains logic for creating nodes and edges that act as "gadgets" in CredRank.
+ * They are most directly used by markovProcessGraph.js
+ */
+
+import {type NodeAddressT, NodeAddress} from "../graph";
+import type {TimestampMs} from "../../util/timestamp";
+import type {MarkovNode} from "./markovNode";
+
+export const CORE_NODE_PREFIX: NodeAddressT = NodeAddress.fromParts([
+  "sourcecred",
+  "core",
+]);
+
+export interface NodeGadget<T> {
+  // Prefix shared by all nodes generated by this gadget.
+  prefix: NodeAddressT;
+  // Convert a "structured address" for nodes of this gadget into raw node addresses.
+  toRaw: (T) => NodeAddressT;
+  // Convert an address generated by this gadget into a structured address. Should
+  // error if the address was not generated by this gadget.
+  fromRaw(NodeAddressT): T;
+  // Generate a full node for this gadget from the structured address.
+  node: (T) => MarkovNode;
+}
+
+export const seedGadget: NodeGadget<void> = (() => {
+  const description: string = "\u{1f331}"; // U+1F331 SEEDLING
+  const prefix = NodeAddress.append(CORE_NODE_PREFIX, "SEED");
+  const toRaw = () => prefix;
+  const fromRaw = () => {};
+  const node = () => ({address: prefix, description, mint: 0});
+  return Object.freeze({prefix, toRaw, fromRaw, node});
+})();
+
+export type EpochAccumulatorAddress = {|
+  +epochStart: TimestampMs,
+|};
+export const accumulatorGadget: NodeGadget<EpochAccumulatorAddress> = (() => {
+  const prefix = NodeAddress.append(CORE_NODE_PREFIX, "EPOCH_ACCUMULATOR");
+  function toRaw(addr) {
+    return NodeAddress.append(this.prefix, String(addr.epochStart));
+  }
+  function fromRaw(addr) {
+    if (!NodeAddress.hasPrefix(addr, this.prefix)) {
+      throw new Error(
+        "Not an epoch node address: " + NodeAddress.toString(addr)
+      );
+    }
+    const prefixLength = NodeAddress.toParts(prefix).length;
+    const parts = NodeAddress.toParts(addr);
+    const epochStart = +parts[prefixLength];
```

Comment:
To be precise here, you would check that `String(epochStart)` matches
`parts[prefixLength]` and fail otherwise. As written, I can give an
address whose last part is `2.00000`, which will parse just the same as
if it were `2`, even though that address did not come from this gadget.

That is, you want to draw a bijection between `T` and a subobject of
`NodeAddressT`.

Similarly, fail if `parts.length` is too long.